### DISCUSSION TITLE
Problem with sensitivity from point-source IRFs

### DISCUSSION THREAD

**Discussion by kosack** [gammapy/gammapy#5417] [Help]
This is following a conversation started in Slack by @tobychev.  

*  **Goal**: Produce a sensitivity curve with gammapy from an existing point-like IRF file
* **Issue:** the resulting sensitivity curve contains no background rate, and therefore incorrect sensitivity

The problem seems to arise in SpectrumMaker, which returns all zeros for the background rate (however it works fine for a full-enclosure IRF).  The issue appears to be with the solid angle applied in `gammapy.makers.utils.make_map_background_irf`, which is 0 steradians (so clearly not using the RAD_MAX).  

The tutorial mentions specifically that a `PointSkyRegion` should be used, which signals that the `SpectrumMaker` should use the _RAD_MAX_ information in the IRF.  However, this doesn't seem to be the case. 

>If a geometry based on a [PointSkyRegion](https://astropy-r Regions.readthedocs.io/en/latest/api/regions.PointSkyRegion.html#regions.PointSkyRegion) is fed to the spectra and the background Makers, the latter will automatically use the values stored in the RAD_MAX keyword / table for defining the size of the ON and OFF regions.

Is there another way to make a energy-dependent ON region perhaps?

A complete example of this can be found here: https://github.com/Tobychev/GammapySensitivity/blob/main/simple_test.ipynb, along with test data if you clone the repo. 

The question is: is this a bug or unsupported feature in gammapy? Or is the method used to compute the sensitivity wrong? 

### supporting info:

```py
spectrum_maker = SpectrumDatasetMaker(
    selection=["exposure", "background","edisp"],
    containment_correction=False,
)
dataset = spectrum_maker.run(empty_dataset, obs)
```

The resulting dataset has correct exposure computed, but 0 background rate. 
```
SpectrumDataset
---------------

  Name                            : 4GNzacBO 

  Total counts                    : 0 
  Total background counts         : 0.00
  Total excess counts             : 0.00

  Predicted counts                : 0.00
  Predicted background counts     : 0.00
  Predicted excess counts         : nan

  Exposure min                    : 5.89e+05 m2 s
  Exposure max                    : 8.81e+11 m2 s

  Number of total bins            : 22 
  Number of fit bins              : 22 

  Fit statistic type              : cash
  Fit statistic value (-2 log(L)) : nan

  Number of models                : 0 
  Number of parameters            : 0
  Number of free parameters       : 0

```

**Comment by kosack** [comment]
Reply in Slack from @adonath:

> In principle there is a unit test for the 2D background case: https://github.com/gammapy/gammapy/blob/67ee75edc95a538b8dd6b6d6f188de2ca6daf0c2/gammapy/makers/tests/test_utils.py#L179
However the point like IRFs with energy dependent rad max are special cased with a point like region, which misses the information on the solid angle of the region. Usually the background is derived from off data in this case, which is handled as a special case in the background makers. When those are not used, the (default) background integration might just return zero in this case, because the solid angle is zero. Yes, I guess we should fix this in Gammapy. I have proposed an energy dependent RegionGeom for this, but we did not have the resources to implement it. So handling this as another special case for now might be fine.

So it seems this is realy a bug or missing feature in gammapy, not a misunderstanding of how to use the code. The decision is to either:
1.  look into implementing an `EnergyDependentCircleSkyRegion`
2. Modify he background code to compute the solid angle for a `PointSkyRegion` using the `RAD_MAX` table instead of returning zero, if a RAD_MAX table exists.

**Reply by kosack** [reply]
I'm inclined to think option 2 is easiest, since to implement 1, it probably requires changing the API of _astropy.regions_ such that energy is somehow passed inside or alongside a SkyCoord. That sounds like too large a change. But maybe experts can see an easier way.

**Reply by adonath** [reply]
Yes, I agree option 2 is the easier solution for now. This is what we did for the reflected region background estimation as well. 

The long-term solution for Gammpy is then an energy dependent `RegionGeom`, which requires a bit more work. Actually I just remembered that I looked into this a few years back and already had a working prototype. The main issue was that `regions` does not support array valued regions (for no good reason). So I had to work around this from the Gammapy side:

- Here is a notebook demonstrating the functionality: https://github.com/adonath/notebooks-public/blob/master/region-geom-array.ipynb
- And the corresponding branch: https://github.com/adonath/gammapy/tree/region_geom_array

I think the core functionality was not that hard, but there was a lot of missing work related to serialization.

**Reply by Tobychev** [reply]
I didn't really understand how to use @adonath's code in my sensitivity calculation, so I ended up writing a `RegionGeom` subclass that takes a rad_max IRF at creation time and then needs a further setup call but otherwise serves as a drop in replacement for `RegionGeom` in the sensitivity calculation. 

I've tested it and it seems to work quite well: 
![Sens](https://github.com/user-attachments/assets/49d17d03-ac76-4260-bb41-4fd665238392)

This is the class:
```python
class RadMaxGeom(RegionGeom):

    def __init__(self, region, rad_max, radii=None, pointing=None, axes=None, wcs=None, binsz_wcs="0.1 deg"):
        covering_region = CircleSkyRegion(center=region.center, radius=rad_max.data.max() * rad_max.unit)
        super().__init__(covering_region, axes=axes, wcs=wcs, binsz_wcs=binsz_wcs)

        if isinstance(rad_max, RadMax2D):
            self._rad_max = rad_max
        else:
            raise ValueError(f"rad_max needs to be instance of `RadMax2d`, got {type(rad_max)}")

        self._radii = radii
        self._pointing = pointing


    @classmethod
    def create(cls, region, rad_max, **kwargs):
        return cls.from_regions(regions=region, rad_max=rad_max, **kwargs)


    def setup_for_solid_angle(self, pointing, obs):
        point_icrs = pointing.get_icrs(obs.tmid)
        reg_coord = self.to_image().get_coord()
        sky_coord = reg_coord.skycoord
        offset = sky_coord.separation(point_icrs)
        self._pointing = point_icrs
        if self._rad_max.axes["offset"].nbin > 1:
            self._radii = np.squeeze(self._rad_max.evaluate(offset=offset, energy=self.axes["energy"].center))
        else:
            self._radii = np.squeeze(self._rad_max.evaluate())

    def solid_angle(self):
        if self.region is None:
            raise ValueError("Region definition required.")

        areas = np.zeros_like(self._radii.value)
        for idx, r in enumerate(self._radii):
            area = CircleSkyRegion(self._pointing, radius=r).to_pixel(self.wcs).area
            areas[idx] = area

        d_omegas = area * proj_plane_pixel_area(self.wcs) * u.deg**2
        return d_omegas.to("sr")
```

**Reply by adonath** [reply]
Thanks @Tobychev,  this look like a much better agreement! Your implementation does essentially the same thing, except that you use loops over regions, while my suggested solution uses arrays and broadcasting (which is the right approach). But I can see that the abstraction level is higher in my approach. From an API perspective a new `RegionGeom` class does not really make sense, because we have the equivalent concept (pixel size changing with energy) already for `WcsGeom`. I would love to see the work started in https://github.com/gammapy/gammapy/pull/5420 to be continued. Let me know if you have some availability, then I am happy to assist in making this a contribution to Gammapy. Otherwise I think you can still continue your work on sensitivity with the proposed work-around.

**Reply by Tobychev** [reply]
Hi @adonath, I gave it a go and came up with the a patch to #5420  that frankly produces even better agreement than my class above as seen here:
![better_sens](https://github.com/user-attachments/assets/db0345a5-28eb-416c-9b92-1637dee42218)

I don't know a good way to push changes to a PR that comes from a fork without also cloning the fork, so I just provide the actual patch below. Notworthy canges is a rather ugly `area` function that mostly just adds two extra axes to line up with the dimensions from `integrate_log_log`, and sneaking in an `to_image` call into the `RegionGeom.contains`. Also had to rename some region things, I guess they've changed names since you first wrote it.

```
diff --git a/gammapy/maps/region/geom.py b/gammapy/maps/region/geom.py
index afe482a48..066f4049f 100644
--- a/gammapy/maps/region/geom.py
+++ b/gammapy/maps/region/geom.py
@@ -65,7 +65,6 @@ class RegionGeom(Geom):
         is provided, the input of ``binsz_wcs`` is overridden.
     """
 
-    is_regular = True
     is_allsky = False
     is_hpx = False
     is_region = True
@@ -249,7 +248,7 @@ class RegionGeom(Geom):
             return np.ones(coords.skycoord.shape, dtype=bool)
 
         if not self.is_regular:
-            idx = self.axes.coord_to_idx(coords)
+            idx = self.to_image().axes.coord_to_idx(coords)
             cls = self.region.__class__
             region = cls(
                 center=self.region.center, radius=self.region.radius[idx].squeeze()
diff --git a/gammapy/utils/regions.py b/gammapy/utils/regions.py
index bd4a8c836..48282900b 100644
--- a/gammapy/utils/regions.py
+++ b/gammapy/utils/regions.py
@@ -12,6 +12,7 @@ TODO: before Gammapy v1.0, discuss what to do about ``gammapy.utils.regions``.
 Options: keep as-is, hide from the docs, or to remove it completely
 (if the functionality is available in ``astropy-regions`` directly.
 """
+
 import operator
 
 import numpy as np
@@ -27,8 +28,8 @@ from regions import (
     Regions,
 )
 from regions._utils import pixel_scale_angle_at_skycoord
-from regions.core import BoundingBox
-from regions.core.attributes import RegionAttr
+from regions.core import RegionBoundingBox
+from regions.core.attributes import RegionAttribute
 from scipy.optimize import Bounds, minimize
 
 __all__ = [
@@ -284,7 +285,7 @@ def region_circle_to_ellipse(region):
 # This code is only needed because regions forbids array valued attributes
 
 
-class ArrayQuantityLength(RegionAttr):
+class ArrayQuantityLength(RegionAttribute):
     """
     Descriptor class for `~regions.SkyRegion`, which takes a scalar
     `~astropy.units.Quantity` object.
@@ -295,7 +296,7 @@ class ArrayQuantityLength(RegionAttr):
             raise ValueError(f"The {self.name} must be an astropy  `Quantity object`")
 
 
-class ArrayLength(RegionAttr):
+class ArrayLength(RegionAttribute):
     """
     Descriptor class for `~regions.PixelRegion`, which takes a scalar
     python/numpy number.
@@ -323,10 +324,15 @@ class CirclePixelRegionArray(CirclePixelRegion):
 
     @property
     def bounding_box(self):
-        """Bounding box (`~regions.BoundingBox`)."""
+        """Bounding box (`~regions.RegionBoundingBox`)."""
         radius = max(self.radius)
         xmin = self.center.x - radius
         xmax = self.center.x + radius
         ymin = self.center.y - radius
         ymax = self.center.y + radius
-        return BoundingBox.from_float(xmin, xmax, ymin, ymax)
+        return RegionBoundingBox.from_float(xmin, xmax, ymin, ymax)
+
+    @property
+    def area(self):
+        # Extra axes to match dimensions from a RegionGeom
+        return super().area[:, np.newaxis, np.newaxis]
```

**Reply by Tobychev** [reply]
There was a bug in the class above that meant only one on-region area was being used, fixing it both implementations give the same result:
![good_sens](https://github.com/user-attachments/assets/1f9be45c-fa63-4902-9fa3-64de40c852df)

**Comment by registerrier** [comment]
A side question is what should one expect as background IRF for RAD_MAX IRFs. 

Is it non integrated bkg brightness in TeV-1 s-1 sr-1? Or should it be count rate in TeV-1 s-1, i.e. already integrated over the correct solid angle?

**Reply by kosack** [reply]
Good question. I would guess already integrated would be the least error prone, but of course if you have the rate/ster and the bin-wise solid angle of the region, it's easy to convert.   

The real goal here was to ensure that if we produce a sensitivity  curve the traditional way starting from events, that gammapy can reproduce it (at least in bins with low energy bias) from IRFs, as a sort of round-trip cross check. In the event-wise case, we compute the integral background rate per bin, but it shouldn't matter much.

**Reply by adonath** [reply]
I don't really have an opinion here. I think it is mostly relevant for sensitivity estimation. For real analyses you either have the full "3d" background or you need to get from off regions anyway. So one could just stick with the full 3d background and do the integration during Gammapy data reduction. As  far as I know there is no specification of this kind of "point like" background in GADF.

**Reply by Tobychev** [reply]
There is an (implicit) specification for point-like background IRFs in GADF, it reads

> The format of each point-like IRF component is analog to the ones already described within the full enclosure IRF specifications (see [Full-enclosure IRFs](https://gamma-astro-data-formats.readthedocs.io/en/latest/irfs/full_enclosure/index.html#full-enclosure-irfs)), with certain differences listed in this section.

which isn't terribly precise, but could reasonably be read to mean it should have the same units as the full enclosure BKG irf: "s^-1 MeV^-1 sr^-1".

**Reply by kosack** [reply]
maybe a more fundamental question is if we should support point-like IRFs at all? If nobody is using them, is there a good reason? Suporting E-dep On regions may be enough - if that gives the same sensitivity as the public IRFs, then there is no good reason to support point-like.  They were really more of a hack in the past to be able to do point-source analysis without knowing the PSF, but if gammapy can correct for PSF overspill, perhaps it's just added complexity to support two different types of IRF.

**Reply by bkhelifi** [reply]
Hi Karl,
I thought about that... But I am better sure that extragal groups would like to have point-like IRFs. First because of a dedicated point-source optimization, because of the very important energy bias at the lowest energy that is correleted with the PSF (aspect that one can not recover with the full-containment IRFs). Maybe a clear statement of the science WG might be needed for that...
The confirmation or the abandon of type of IRF should end by a requirement for DPPS/SUSS and the sc operation, and then on the data format (as mentioned @registerrier)
